### PR TITLE
Moving Expense Document Model / Swagger

### PR DESCRIPTION
**Pull request by macrael**
## Description

Adds a model and swagger definitions for Moving Expense Document which provides metadata for uploads related to moving expense. 

## Reviewer Notes

Consider the consequences of using the Reimbursement model like this.

## Code Review Verification Steps

* [ ] All tests pass.
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Any migrations/schema changes also update the diagram in docs/schema/dp3.sqs
* [ ] This works in IE.
* [ ] Request review from a member of a different team.
* [ ] (TRIAL) Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/159074788) for this change

### CODE REVIEW

**Comment by macrael** [comment]
I still need to add the model diagram update. will have to wait until after my interview.

File: swagger/internal.yaml [hunk 1]
```diff
@@ -2478,6 +2561,62 @@ paths:
           description: not authorized to modify this move
         500:
           description: server error
+  /moves/{moveId}/moving_expense_documents:
+    get:
+      summary: Returns a list of all Moving Expense Documents associated with this move
+      description: Returns a list of all Moving Expense Documents associated with this move
+      operationId: indexMovingExpenseDocuments
+      tags:
+        - moving_expenses
```
Comment (stangah) [inline swagger/internal.yaml]: We tagged move document routes with `moves`, probably want to continue that here, or maybe make an encompassing `move_documents` tag for all the routes we'll need

File: swagger/internal.yaml [hunk 2]
```diff
@@ -2478,6 +2561,62 @@ paths:
           description: not authorized to modify this move
         500:
           description: server error
+  /moves/{moveId}/moving_expense_documents:
```
Comment (tinyels) [inline swagger/internal.yaml]: Is it possible to have the move documents endpoint be polymorphic? That is, can it handle move documents and moving expenses? (Open API 3.0 has a `oneOf` keyword that would make that possible but I am not sure about the version we are using.)

Comment (jim) [reply swagger/internal.yaml]: We're on 2.0, which [doesn't have `oneOf`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) (at least that I can find).

Comment (macrael) [reply swagger/internal.yaml]: We are using 2.0

Comment (macrael) [reply swagger/internal.yaml]: I'll look into what we can do with 2.0, right now it's it's own endpoint

File: pkg/models/moving_expense_document.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	MovingExpenseTypeOIL MovingExpenseType = "OIL"
+	// MovingExpenseTypeOTHER captures enum value "OTHER"
+	MovingExpenseTypeOTHER MovingExpenseType = "OTHER"
+)
+
+// MovingExpenseDocument is an object representing a move document
+type MovingExpenseDocument struct {
+	ID                uuid.UUID         `json:"id" db:"id"`
+	MoveDocumentID    uuid.UUID         `json:"document_id" db:"document_id"`
+	MoveDocument      MoveDocument      `belongs_to:"documents"`
```
Comment (amitch23) [inline pkg/models/moving_expense_document.go]: Should this belong to move_documents instead of documents?

File: swagger/internal.yaml [hunk 4]
```diff
@@ -510,6 +510,43 @@ definitions:
     type: array
     items:
       $ref: '#/definitions/MoveDocumentPayload'
+  GenericMoveDocumentPayload:
```
Comment (amitch23) [inline swagger/internal.yaml]: I don't see where this payload is being used - is it supposed to be?

File: pkg/models/moving_expense_document.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	MovingExpenseTypeOIL MovingExpenseType = "OIL"
+	// MovingExpenseTypeOTHER captures enum value "OTHER"
+	MovingExpenseTypeOTHER MovingExpenseType = "OTHER"
+)
+
+// MovingExpenseDocument is an object representing a move document
+type MovingExpenseDocument struct {
+	ID                uuid.UUID         `json:"id" db:"id"`
+	MoveDocumentID    uuid.UUID         `json:"move_document_id" db:"move_document_id"`
+	MoveDocument      MoveDocument      `belongs_to:"documents"`
```
Comment (amitch23) [inline pkg/models/moving_expense_document.go]: Should the MoveDocument belong to 'move_documents'?

File: swagger/internal.yaml [hunk 6]
```diff
@@ -597,11 +631,137 @@ definitions:
...
+      id:
+        type: string
+        format: uuid
+        example: c56a4180-65aa-42ec-a945-5fd21dec0538
+      move_id:
+        type: string
+        format: uuid
+        example: c56a4180-65aa-42ec-a945-5fd21dec0538
+      document:
+        $ref: '#/definitions/DocumentPayload'
```
Comment (amitch23) [inline swagger/internal.yaml]: Should this be a reference to move document, not document?

File: swagger/internal.yaml [hunk 7]
```diff
@@ -2647,6 +2810,78 @@ paths:
           description: move document not found
         500:
           description: internal server error
+  /moves/{moveId}/moving_expense_documents:
+    post:
```
Comment (amitch23) [inline swagger/internal.yaml]: Can you add an `index` here?

File: swagger/internal.yaml [hunk 8]
```diff
@@ -597,11 +634,137 @@ definitions:
...
+    required:
+      - upload_ids
+      - title
+      - move_document_type
+      - moving_expense_type
+      - reimbursement
+  UpdateMovingExpenseDocumentPayload:
+    type: object
+    properties:
+      id:
```
Comment (stangah) [inline swagger/internal.yaml]: We don't yet need to support adding _more_ uploads to a move document, right?

File: swagger/internal.yaml [hunk 9]
```diff
@@ -2647,6 +2810,78 @@ paths:
...
+            $ref: '#/definitions/MovingExpenseDocumentPayload'
+        400:
+          description: invalid request
+        401:
+          description: must be authenticated to use this endpoint
+        403:
+          description: not authorized to modify this move
+        500:
+          description: server error
+  /moves/{moveId}/moving_expense_documents/{movingExpenseDocumentId}:
```
Comment (stangah) [inline swagger/internal.yaml]: Does the update need to be nested under the move route?

File: swagger/internal.yaml [hunk 10]
```diff
@@ -597,11 +634,137 @@ definitions:
       - WEIGHT_TICKET
       - STORAGE_EXPENSE
       - SHIPMENT_SUMMARY
+      - EXPENSE
     x-display-value:
       OTHER: Other document type
       WEIGHT_TICKET: Weight ticket
       STORAGE_EXPENSE: Storage expense receipt
       SHIPMENT_SUMMARY: Shipment summary
+  MovingExpenseDocumentPayload:
```
Comment (stangah) [inline swagger/internal.yaml]: So is the pattern going forward that, for these subtypes of `MoveDocument`, we unnest the fields into a single payload and kind of hide the fact that they're two models from the client?

Comment (macrael) [reply swagger/internal.yaml]: yes, we want them to look like the same object field wise, with different types.

File: swagger/internal.yaml [hunk 11]
```diff
@@ -2647,6 +2810,78 @@ paths:
...
+        - in: body
+          name: updateMovingExpenseDocument
+          required: true
+          schema:
+            $ref: '#/definitions/UpdateMovingExpenseDocumentPayload'
+      responses:
+        200:
+          description: updated instance of moving expense document
+          schema:
+            $ref: '#/definitions/MoveDocumentPayload'
```
[the remark this answers is not in the document]
Comment (amitch23) [reply swagger/internal.yaml]: done

**Pull request merged**